Add support for explicit user search by username and email in Admin UI

thomasdarimont opened the pull request:
This adds support for the `username:$USERNAME` `email:$EMAIL` in the vein of `id:$USERID`.

Add support for user lookup by username
- Add ability for direct user lookups by username and email to UsersResource and BruteForceUsersResource
- Document user search by selected fields
- Extract SEARCH_ID_PARAMETER to SearchQueryUtils

Fixes https://github.com/keycloak/keycloak/issues/26602

Discussion: #13184

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

Nowheresly commented:
really useful feature
+1

thomasdarimont commented:
The tests work fine on my side. It seems as if the keycloak-documentation build has some issues at the moment.

pedroigor commented:
@mposolda @ssilvert Wdyt?

pedroigor commented:
I don't see why we can't have e-mail in this PR?

thomasdarimont commented:
Okay then I'll add email too for completeness

ssilvert commented:
@thomasdarimont We already added user attribute searches to the UI.  I take it that you are saying these searches are too inefficient?

thomasdarimont commented:
@ssilvert thanks for having a look. I think you are referring to this regarding attribute search: https://github.com/keycloak/keycloak/blob/main/rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/BruteForceUsersResource.java#L97C24-L97C37
This method will internally call https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L733. This in turn will generate a generic search predicate: https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L900
This method generates something like `... username = ? OR email = ? OR lower(firstname) = ? OR lower(lastname) = ? ....`.

This of course does the job but is not very efficient. I think the proposed shortcuts for very common search operations speed up user search via admin-ui and rest API significantly.

thomasdarimont committed:
Generalize user search prefix lookups via UserSearchPrefix enum (#26602)

- Introduce UserSearchPrefix enum in SearchQueryUtils pairing each prefix
(id:, username:, email:) with its UserProvider lookup,
plus a splitTerms helper backed by a precompiled "\\s+" pattern
- Collapse duplicate prefix branches in UsersResource#getUsers,
UsersResource#getUsersCount and BruteForceUsersResource#searchUser
- BruteForceUsersResource: support multi-term lookups (e.g. "username:foo bar"),
aligning with UsersResource
- Tests: add searchByUsernameSearch / searchByEmailSearch covering
single-term, multi-term and whitespace-tolerant variants
- Docs: add "Search by fields" section to proc-searching-user.adoc

Fixes #26602

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
I rebased my branch on current main and reimplemented the support for explicit user and email lookups via `username:...` and `email:...` prefixes. 
I also aligned the lookup used by UsersResource and BruteForceProtected users.

pedroigor reviewed (changes requested):
Thanks, @thomasdarimont.

Can you please update the PR:

* Add `email` to `getCount`
* Add tests to cover `count` within the test methods you are adding

Once we have that, LGTM.

thomasdarimont commented:
Just pushed a rebased version

thomasdarimont commented:
@pedroigor I think searches for prefixes like id:, email: and username: are covered now. I added some additional checks for count(..) as you requested.
